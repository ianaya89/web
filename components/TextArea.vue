<template>
  <div class="relative">
    <span v-if="soloLabel && outline" class="text-grey text-p2 uppercase tracking-wide">
      {{ soloLabel }}
    </span>
    <span v-if="soloLabel && required && outline" class="text-red text-p2">
      *
    </span>
    <v-textarea
      :no-resize="noResizeProp"
      :single-line="stickyLabel"
      v-bind="$props"
      :class="[requiredClass]"
      class="textarea"
      :error-messages="translatableErrorMessages"
      @change="onChange"
      @keyup.enter="$emit('enterClick')"
      @input="onInput"
    />
  </div>
</template>

<script>
import BaseInput from '@/components/BaseInput'
import SingleUpload from '@/components/SingleUpload'

export default BaseInput.extend({
  name: 'TextArea',

  components: {
    SingleUpload
  },

  props: {
    name: {
      type: String,
      default: 'name'
    },
    autoGrow: {
      type: Boolean,
      default: false
    },
    counter: {
      type: Number,
      default: null
    },
    noResizeProp: {
      type: Boolean,
      default: false
    },
    rows: {
      type: Number,
      default: 5
    },
    soloLabel: {
      type: String,
      default: 'Label'
    },
    errorMessages: {
      type: Array,
      default() {
        return []
      }
    }
  },

  computed: {
    requiredClass() {
      return this.required ? 'required-sign' : 'no-required'
    },

    translatableErrorMessages() {
      return this.errorMessages.map(message => this.$t(message))
    }
  },

  methods: {
    onChange(value) {
      this.onInput(value)
      this.$emit('change', value)
    },
    onInput(value) {
      this.$emit('input', value)
    }
  }
})
</script>

<style lang="postcss">
.v-textarea.v-text-field--enclosed textarea {
  margin-top: 0px;
}

.textarea {
  textarea {
    @apply .min-h-20 .leading-tight;
  }

  .v-textarea textarea {
    flex: 1 1 auto;
    line-height: 18px;
  }

  &.required-sign {
    label::after {
      @apply text-red;
      content: '*';
    }
  }

  &.no-required {
  }
}

.v-text-field.v-text-field--solo > .v-input__control > .v-input__slot {
  box-shadow: 0px 3px 1px -2px rgba(0, 0, 0, 0) !important;
}
</style>
