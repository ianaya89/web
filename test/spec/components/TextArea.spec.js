import { mount, shallowMount } from '@vue/test-utils'
import TextArea from '@/components/TextArea'

describe('TextArea', () => {
  it('renders default textarea', () => {
    const wrapper = shallowMount(TextArea)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders sticky label', () => {
    const wrapper = shallowMount(TextArea, {
      propsData: {
        stickyLabel: true
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders input with placeholder outline and hint', () => {
    const wrapper = shallowMount(TextArea, {
      propsData: {
        placeholder: 'Your name',
        outline: true,
        label: 'First name',
        hint: 'Please write your first name',
        persistentHint: true
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('check input value', () => {
    const wrapper = mount(TextArea, {
      propsData: {
        value: 'Hello'
      }
    })

    const textarea = wrapper.find('textarea')
    expect(textarea.element.value).toBe('Hello')
    textarea.setValue('value')
    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toEqual('value')
  })
})
