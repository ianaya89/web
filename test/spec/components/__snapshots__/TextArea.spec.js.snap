// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders default textarea 1`] = `
<div
  class="relative"
>
  <!---->
   
  <!---->
   
  <v-textarea-stub
    backgroundcolor=""
    class="textarea no-required"
    clearicon="$vuetify.icons.clear"
    color="primary"
    errorcount="1"
    errormessages=""
    label="Label"
    messages=""
    name="name"
    rowheight="24"
    rows="5"
    rules=""
    sololabel="Label"
    successmessages=""
    type="text"
  />
</div>
`;

exports[`TextArea renders input with placeholder outline and hint 1`] = `
<div
  class="relative"
>
  <span
    class="text-grey text-p2 uppercase tracking-wide"
  >
    
    Label
  
  </span>
   
  <!---->
   
  <v-textarea-stub
    backgroundcolor=""
    class="textarea no-required"
    clearicon="$vuetify.icons.clear"
    color="primary"
    errorcount="1"
    errormessages=""
    hint="Please write your first name"
    label="First name"
    messages=""
    name="name"
    outline="true"
    persistenthint="true"
    placeholder="Your name"
    rowheight="24"
    rows="5"
    rules=""
    sololabel="Label"
    successmessages=""
    type="text"
  />
</div>
`;

exports[`TextArea renders sticky label 1`] = `
<div
  class="relative"
>
  <!---->
   
  <!---->
   
  <v-textarea-stub
    backgroundcolor=""
    class="textarea no-required"
    clearicon="$vuetify.icons.clear"
    color="primary"
    errorcount="1"
    errormessages=""
    label="Label"
    messages=""
    name="name"
    rowheight="24"
    rows="5"
    rules=""
    singleline="true"
    sololabel="Label"
    stickylabel="true"
    successmessages=""
    type="text"
  />
</div>
`;
